### PR TITLE
Use HTTPS links where possible

### PR DESCRIPTION
**Pull request by syvb**
While looking at #86583, I wondered how many other (insecure) HTTP links were in `rustc`. This changes most other `http` links to `https`. While most of the links are in comments or documentation, there are a few other HTTP links that are used by CI that are changed to HTTPS.

Notes:
- I didn't change any to or in licences
- Some links don't support HTTPS :(
- Some `http` links were dead, in those cases I upgraded them to their new places (all of which used HTTPS)

### CODE REVIEW

File: compiler/rustc_target/src/abi/call/msp430.rs [hunk 1]
```diff
@@ -1,5 +1,5 @@
 // Reference: MSP430 Embedded Application Binary Interface
-// http://www.ti.com/lit/an/slaa534/slaa534.pdf
+// https://web.archive.org/web/20191221205752/http://www.ti.com/lit/an/slaa534/slaa534.pdf
```
Comment (petrochenkov) [inline compiler/rustc_target/src/abi/call/msp430.rs:2]: https://www.ti.com/lit/an/slaa534a/slaa534a.pdf seems to be the alive link.

File: src/test/ui/nullable-pointer-iotareduction.rs [hunk 2]
```diff
@@ -4,7 +4,7 @@
 
 // Iota-reduction is a rule in the Calculus of (Co-)Inductive Constructions,
 // which "says that a destructor applied to an object built from a constructor
-// behaves as expected".  -- http://coq.inria.fr/doc/Reference-Manual006.html
+// behaves as expected".  -- https://web.archive.org/web/20100531091244/http://coq.inria.fr/doc/Reference-Manual006.html
```
Comment (petrochenkov) [inline src/test/ui/nullable-pointer-iotareduction.rs:7]: https://coq.inria.fr/doc/language/core/conversion.html#iota-reduction

**Comment by petrochenkov** [comment]
LGTM, although I obviously didn't check whether all the links are accessible.

**Comment by petrochenkov** [comment]
CI is unable to download from https://gcc.gnu.org/pub/gcc/infrastructure/mpfr-2.4.2.tar.bz2
```
--2021-06-25 04:05:16--  https://gcc.gnu.org/pub/gcc/infrastructure/mpfr-2.4.2.tar.bz2
Resolving gcc.gnu.org... 8.43.85.97, 2620:52:3:1:0:246e:9693:128c
Connecting to gcc.gnu.org|8.43.85.97|:443... connected.
OpenSSL: error:1407742E:SSL routines:SSL23_GET_SERVER_HELLO:tlsv1 alert protocol version
Unable to establish SSL connection.
```

**Comment by dns2utf8** [comment]
On the mailinglist from 2015 there was a discussion about including TLS 1.2 in Debian 6 (the image we are using):
https://lists.debian.org/debian-lts/2015/02/msg00012.html

It should support https because it is able to download from AWS.

The host gcc.gnu.org supports TLSv1.2 only:
https://www.ssllabs.com/ssltest/analyze.html?d=gcc.gnu.org&s=8.43.85.97&hideResults=on&ignoreMismatch=on

Where AWS supports TLS 1.0 to 1.3 the old image can use 1.1.

Should we update it to Debian 7 because TLS 1.2 support will not happen and the LTS support has ended in 2016?

**Comment by petrochenkov** [comment]
>Should we update it to Debian 7 because TLS 1.2 support will not happen and the LTS support has ended in 2016?

Maybe, but certainly not in this PR.
If changing http to https doesn't "just work" and require infra tweaks, then those changes and tweaks need to go to a separate PR (probably assigned to Mark-Simulacrum).

**Pull request merged**